docs/tutorials/10_effective_dimension.ipynb

```diff
@@ -0,0 +1,692 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# we can set the total number of input samples and weight samples for random selection\n",
+    "num_input_samples = 10\n",
+    "num_weight_samples = 10\n",
+    "\n",
+    "global_ed = EffectiveDimension(\n",
+    "    qnn=qnn, weight_samples=num_weight_samples, input_samples=num_input_samples, callback=callback\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If we want to test a specific set of input samples and weight samples, we can provide it directly to the `EffectiveDimension` class as shown in the following snippet:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# we can also provide user-defined samples and parameters\n",
+    "input_samples = algorithm_globals.random.normal(0, 1, size=(10, qnn.num_inputs))\n",
+    "weight_samples = algorithm_globals.random.uniform(0, 1, size=(10, qnn.num_weights))\n",
+    "\n",
+    "global_ed = EffectiveDimension(\n",
+    "    qnn=qnn, weight_samples=weight_samples, input_samples=input_samples, callback=callback\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The effective dimension algorithm also requires a dataset size. In this example, we will define an array of sizes to later see how this input affects the result."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# finally, we will define ranges to test different numbers of data, n\n",
+    "n = [5000, 8000, 10000, 40000, 60000, 100000, 150000, 200000, 500000, 1000000]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.2 Compute the Global Effective Dimension\n",
+    "Let's now calculate the effective dimension of our network for the previously defined set of input samples, weights, and a dataset size of 5000. The previously defined callback function will show the time each pass takes in the Monte Carlo simulation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Iteration: 0, forward time: 0.03782296180725098, backward time: 0.36411023139953613\n",
+      "Iteration: 1, forward time: 0.01554107666015625, backward time: 0.45325207710266113\n",
+      "Iteration: 2, forward time: 0.01328730583190918, backward time: 0.2095649242401123\n",
+      "Iteration: 3, forward time: 0.014077186584472656, backward time: 0.25399303436279297\n",
```

Comment:
I am wondering about the value of this effective dimension callback. In other callbacks elsewhere we have things like current expectation value, current loss value as these are computed at each step that these might be used by the user to track/plot algorithm progression etc. Other information that the user might use is also passed, such as current job for execution, that they can then user other calls to look up the job and see more detail etc. Timing information like this I would at first have thought more think about logging, so if they wanted it they would enable logging . I guess  each "iteration" here, where there are 10 of them,  relates to the 10 samples, so its more an index into the sample being processing in order? 

Also as it stands I see the parameters of the callback are not documented in the algorithm constructor - I looked there to see what it said about this.

If its felt that this is desirable to keep, I think in this example I would format the times to say limit them to say 3 decimal places rather than have so many decimal places that I imagine mean little as I imagine these times vary quite a bit from run to run depending what else is going on in the machine running these.

Reply:
The reasoning behind the callback is that iteration times are interesting for monitoring the performance, particularly when ran inside a runtime program, and logging cannot be accessed in runtime programs. 

I see your point with using the name "iterations", each iteration corresponds to a sampled set of weights (there are 10 samples in this examples), so maybe I can substitute "iteration" with "weights sample". I will also apply the limit on decimal places.

Reply:
In these other callbacks normally the user can do something like tune some hyperparameters, change the anstaz and/or optimizer and can see how that affects say the convergence via the callback.

In the case of timing, aside from perhaps being of interest, I don't think there is anything you can do right. If you are computing the effective dimension of a given network it is going to be whatever it is whether local or on the runtime.

If its something you found useful and its something users would want then ok, but...

> Also as it stands I see the parameters of the callback are not documented in the algorithm constructor - I looked there to see what it said about this.

But I think the parameters ought to be documented as to what they are on the class, and that change could be added to this PR. What it has at present is 

```
        callback: Optional[Callable[[int, float, float], None]] = None,
...        
        callback: A callback function for the Monte Carlo sampling.
```
that does not spell things out. If it can say the int is the current weight sample and the floats are the forward and backward times on seconds for evaluating that sample   - or whatever the right wording is to describe this.

Reply:
@woodsp-ibm It's true that the callback itself is not used for anything practical, and while I did find it useful in certain experiments, I now also see how it makes sense to just turn it into a logging message (it adds noise to the documentation). I can later look into using a callback for the runtime program if I find it necessary, but I agree that the class code is cleaner without it. Sorry for going in circles for so long about this (and thanks for you patience). I now have removed the callback and turned into a debug message. I hope it makes sense now!